cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginService.java

```diff
@@ -129,17 +161,38 @@ public void onRemoval(RemovalNotification<ArtifactDescriptor, Instantiators> not
     private final File pluginDir;
 
     private Instantiators(ArtifactDescriptor parentArtifactDescriptor) throws IOException {
-      this.parentClassLoader = artifactRepository.createArtifactClassLoader(parentArtifactDescriptor.getLocation());;
-      this.pluginInstantiatorMap = new HashMap<>();
+      this.parentClassLoader = artifactRepository.createArtifactClassLoader(parentArtifactDescriptor.getLocation());
+      this.pluginInstantiatorMap = new ConcurrentHashMap<>();
       this.pluginDir = DirUtils.createTempDir(stageDir);
     }
 
+    private boolean isArtifactSnapshotChanged(ArtifactDescriptor artifactDescriptor) {
```

Comment:
snapshot is the most common case, but this can also happen if somebody deletes an artifact, then uploads another one.  So lets just name it 'hasArtifactChanged'